FRB Schema for the CHIME Telescope

tabbott36 opened the pull request:
<!-- IMPORTANT!!! Aside from typographical corrections and minor changes, please consider creating an Issue before making a Pull Request. -->

# Description
Added a schema for fast radio bursts (FRBs). I have also included an example json for FRB alerts from the Canadian Hydrogen Intensity Mapping Experiment (CHIME) for each alert type (detection, subsequent, retraction, and update). 

# Questions for GCN Team
- Can I use a Transient Name Server name for the "id" parameter in the Event schema?

# Testing
- I modelled the schema after other observatories' schemas but am open to suggestions regarding testing.

tabbott36 commented:
Seeing many failed checks, is there action I should take on this? Cheers!

lpsinger commented:
> Seeing many failed checks, is there action I should take on this? Cheers!

Yes, there were some missing contributing instructions to configure pre-commit hooks. See https://github.com/nasa-gcn/gcn-schema/pull/218.

dakota002 commented on gcn/notices/chime/frb.schema.json:
```diff
@@ -0,0 +1,78 @@
+{
+  "$id": "https://gcn.nasa.gov/schema/main/gcn/notices/chime_frb/alert.schema.json",
```
```suggestion
  "$id": "https://gcn.nasa.gov/schema/main/gcn/notices/chime/frb.schema.json",
```

tabbott36 commented:
Hi @dakota002, thanks for the help! Yes, I will revert the versioning and rename the example files.

dakota002 commented on gcn/notices/chime/detection.example.json:
```diff
@@ -0,0 +1,24 @@
+  "alert_type": "detection",
+  "trigger_time": "2024-09-18T07:19:10.765268Z",
+  "trigger_time_error": 0.007864319719374001,
+  "id": "427325191",
```
```suggestion
  "id": ["427325191"],
```
This property comes from https://github.com/nasa-gcn/gcn-schema/blob/main/gcn/notices/core/Event.schema.json, and is defined as an array of numbers or strings

dakota002 commented on gcn/notices/chime/detection.example.json:
```diff
@@ -0,0 +1,24 @@
+{
+  "$schema": "https://gcn.nasa.gov/schema/main/gcn/notices/chime/frb.schema.json",
+  "alert_type": "detection",
```
This is not a valid option, the allowed values are:
```json
"allowedValues": [
        "initial",
        "update",
        "retraction"
      ]

```

These are defined here https://github.com/nasa-gcn/gcn-schema/blob/main/gcn/notices/core/Alert.schema.json

Vidushi-GitHub replied:
@tabbott36 we are working on adding subsequent (see https://github.com/nasa-gcn/gcn-schema/pull/221). 
However, initial will like remain same, else we need to take votes from the missions which are already using it.

For validation, as detection and subsequent please add a new property, current alert_type is not consistent.

tabbott36 replied:
Ok, sounds good to me. Thanks, Vidushi!

dakota002 commented on gcn/notices/chime/retraction.example.json:
```diff
@@ -0,0 +1,8 @@
+{
+  "$schema": "https://gcn.nasa.gov/schema/main/gcn/notices/chime/frb.schema.json",
+  "alert_type": "retraction",
+  "id": "FRB 20231013B",
```
```suggestion
  "id": ["FRB 20231013B"],
```

Same reason as previous example

dakota002 commented on gcn/notices/chime/subsequent.example.json:
```diff
@@ -0,0 +1,26 @@
+  "known_source": "FRB 20231013B",
+  "trigger_time": "2024-09-18T07:19:10.765268Z",
+  "trigger_time_error": 0.007864319719374001,
+  "id": "FRB 20240918A",
```
```suggestion
  "id": ["FRB 20240918A"],
```

dakota002 commented on gcn/notices/chime/subsequent.example.json:
```diff
@@ -0,0 +1,26 @@
+{
+  "$schema": "https://gcn.nasa.gov/schema/main/gcn/notices/chime/frb.schema.json",
+  "alert_type": "subsequent",
```
This is also not a valid value for `alert_type`, would `update` make sense here?

dakota002 commented on gcn/notices/chime/update.example.json:
```diff
@@ -0,0 +1,9 @@
+{
+  "$schema": "https://gcn.nasa.gov/schema/main/gcn/notices/chime/frb.schema.json",
+  "alert_type": "update",
+  "id": "FRB 20231013B",
```
```suggestion
  "id": ["FRB 20231013B"],
```

dakota002 commented:
> Hi @dakota002, thanks for the help! Yes, I will revert the versioning and rename the example files.

No problem! I have suggested a few other changes that should also fix your validation checks

dakota002 commented:
@tabbott36 This looks good to me now, can you please squash your commits down to 1? If you have questions on this I can assist you

tabbott36 commented:
Sure thing! I believe it's been squashed to one commit now. Thanks!

dakota002 commented:
@Vidushi-GitHub @jracusin is this good now?

Vidushi-GitHub commented:
> @Vidushi-GitHub @jracusin is this good now?

We still are working on it.

jracusin commented on gcn/notices/chime/frb.schema.json:
```diff
@@ -0,0 +1,88 @@
+      "type": "number",
+      "description": "A machine learning score from 0-1 separating RFI (0) from an astrophysical signal (1)"
+    },
+    "probability": {
```
```suggestion
    "association_probability": {
```

tabbott36 replied:
In our alerts, we use "probability" to distinguish between likely terrestrial signals (0) and likely astrophysical signals (1). Perhaps the best name for this would be something like "astrophysical_probability" or "astro_origin_probability"?

jracusin commented on gcn/notices/chime/frb.schema.json:
```diff
@@ -0,0 +1,88 @@
+    },
+    "sampling_time": {
+      "type": "number",
+      "description": "The time resolution of the real-time FRB search."
```
in what units?

tabbott36 replied:
Thanks for the review, I will correct this to say milliseconds.

Vidushi-GitHub commented on gcn/notices/chime/frb.schema.json:
```diff
@@ -0,0 +1,88 @@
+  "properties": {
+    "alert_kind": {
+      "type": "string",
+      "description": "The kind of alert (detection or subsequent). Similar meaning to alert_type"
```
This is updated in core schema, please utilize it.

Vidushi-GitHub commented on gcn/notices/chime/frb.schema.json:
```diff
@@ -0,0 +1,88 @@
+    },
+    "snr": {
+      "type": "number",
+      "description": "Signal-to-noise ratio of the burst"
```
Is this different from `rate_snr` in core/Statistics.schema.json? 
```suggestion
      "description": "Signal-to-noise ratio of the burst [dimensionless]"
```

tabbott36 replied:
Though they are very similar, I think it would be misleading to call it "rate_snr" in the context of fast radio bursts. In the high energy regime, rates is usually a measure of photon counts / time, but for radio transients, rates are usually a measure of bursts / time. In other words, a "rate_snr" for a fast radio burst might mislead people into thinking there were multiple bursts detected.

jracusin replied:
We appreciate that rate_snr as used in high-energy is different than what you have here, but could you make it `[something]_snr`, `[something]_duration` so that it has a parallel structure to `rate_snr` and `image_snr`?

Vidushi-GitHub commented on gcn/notices/chime/frb.schema.json:
```diff
@@ -0,0 +1,88 @@
+      "type": "number",
+      "description": "Estimated contribution to the dispersion measure from the Galaxy using the YMW2016 model."
+    },
+    "trigger_time_error": {
```
This property has been added in the core/DateTime.schema.json, please utilize it.

Vidushi-GitHub commented on gcn/notices/chime/frb.schema.json:
```diff
@@ -0,0 +1,88 @@
+      "type": "number",
+      "description": "The time resolution of the real-time FRB search."
+    },
+    "bandwidth": {
```
Do you think it's same as `spectrum` present in core/Reporter.schema.json?

tabbott36 replied:
Ah, yes, it looks like the same thing. "bandwidth" here is the frequencies over which the instrument observes, so I think it alligns well with the "spectrum" property. I'll update it! Thanks

tabbott36 replied:
Looking back at this, I am noticing that "spectrum" is one of ["energy", "wavelength", "frequency"], which I originially misunderstood. Since bandwidth is a number describing the observatory's observing band, I've reintroduced it into the FRB schema.

Vidushi-GitHub commented:
Hi @tabbott36, we have added a few properties in core schema motivated from your CHIME/FRB schema, thanks! 
Please provide your valuable feedback for #226. Please comment on what are usually common properties among FRBs, that should be include in the core.

Vidushi-GitHub commented:
@tabbott36 please use core schema: ``Reporter.schema.json`` and ``DispersionMeasure``

tabbott36 commented:
Hi @Vidushi-GitHub, I've changed the dm to the new DispersionMeasure schema. I also moved the dm_gal_ne_2001_max to the frb schema to avoid conflicts... not sure what the remaining conflict is - do you have an idea on this?

Vidushi-GitHub commented:
> Hi @Vidushi-GitHub, I've changed the dm to the new DispersionMeasure schema. I also moved the dm_gal_ne_2001_max to the frb schema to avoid conflicts... not sure what the remaining conflict is - do you have an idea on this?

I understand it will have merge conflicts, as core schema has changed w.r.t. earlier.
alert_type, trigger_time_error etc are new.
Even for the package.json, reverse to the original. It will be sorted later.

If you are comfortable to create a new PR and acknowledge this PR, that should also work.

tabbott36 commented:
I tried resolving the conflicts and pushing the changes, but still failing the `npm run test`. Getting the error: `Error [ERR_MODULE_NOT_FOUND]: Cannot find package 'ajv'` which is resolved by `npm install ajv
`. Should I go ahead and install `ajv`?

dakota002 commented:
> I tried resolving the conflicts and pushing the changes, but still failing the `npm run test`. Getting the error: `Error [ERR_MODULE_NOT_FOUND]: Cannot find package 'ajv'` which is resolved by `npm install ajv `. Should I go ahead and install `ajv`?

Hi, if you just run `npm i` it will take care of installing all required packages

dakota002 commented on gcn/notices/chime/frb.detection.example.json:
```diff
@@ -0,0 +1,24 @@
+  "dec": 12.632485229956252,
+  "ra_dec_error": [0.503806986334273, 0.598921231135782, 0],
+  "dm": 279.422607421875,
+  "dm_error": 0.40437427163124,
```
```suggestion
  "dm_error": [0.40437427163124],
```
This field needs to be an array

dakota002 commented on gcn/notices/chime/frb.subsequent.example.json:
```diff
@@ -0,0 +1,26 @@
+  "dec": 12.632485229956252,
+  "ra_dec_error": [0.503806986334273, 0.598921231135782, 0],
+  "dm": 279.422607421875,
+  "dm_error": 0.40437427163124,
```
```suggestion
  "dm_error": [0.40437427163124],
```

This field needs to be an array of numbers

Vidushi-GitHub commented on gcn/notices/chime/frb.detection.example.json:
```diff
@@ -0,0 +1,24 @@
+  "snr": 12.698559761047363,
+  "ra": 346.77850859547,
+  "dec": 12.632485229956252,
+  "ra_dec_error": [0.503806986334273, 0.598921231135782, 0],
```
It's fine, [semi-major axis, semi-minor axis, position angle]

Vidushi-GitHub commented:
@tabbott36 we have upgraded our core schema for any error either number or array.
Please pull latest master, it should validate dm_error as number.

Vidushi-GitHub commented on gcn/notices/chime/frb.update.example.json:
```diff
@@ -0,0 +1,9 @@
+  "alert_type": "update",
+  "id": ["FRB 20231013B"],
+  "trigger_time": "2024-09-18T07:19:10.765268Z",
+  "trigger_time_error": [0.007864319719374001],
```
Please make it number rather array.

Vidushi-GitHub replied:
```suggestion
  "trigger_time_error": 0.00786431,
```

Vidushi-GitHub commented on gcn/notices/chime/frb.update.example.json:
```diff
@@ -0,0 +1,9 @@
+  "trigger_time": "2024-09-18T07:19:10.765268Z",
+  "trigger_time_error": [0.007864319719374001],
+  "update_message": "The dispersion measure of FRB 20231013B has been updated to 221.19271850585938 +/- 0.40437427163124 pc/cm^3.",
+  "description": "This alert was generated automatically by human-issued retraction. Please direct comments, questions, and concerns to Thomas Abbott at thomas.abbott@mail.mcgill.ca."
```
Could you please keep the contact info at FRB mission page only?
As Notices will be archival data even after 10 years, but email addresses are temporary.

Vidushi-GitHub commented on gcn/notices/chime/frb.subsequent.example.json:
```diff
@@ -0,0 +1,26 @@
+  "bandwidth": 400,
+  "npol": 2,
+  "tsys": 50,
+  "description": "This alert was generated automatically by the CHIME/FRB Real-time Search Pipeline. Please direct comments, questions, and concerns to Thomas Abbott at thomas.abbott@mail.mcgill.ca."
```
emails are temporary, mission page is appropriate place.

Vidushi-GitHub commented on gcn/notices/chime/frb.retraction.example.json:
```diff
@@ -0,0 +1,8 @@
+  "alert_type": "retraction",
+  "id": ["FRB 20231013B"],
+  "trigger_time": "2024-09-18T07:19:10.765268Z",
+  "trigger_time_error": [0.007864319719374001],
```
```suggestion
  "trigger_time_error": 0.007864319719374001,
```
Now, numbers are also allowed for errors.

Vidushi-GitHub reviewed (approved):
@jracusin looks good, please provide comment/approve it.

Vidushi-GitHub commented on gcn/notices/chime/frb.detection.example.json:
```diff
@@ -0,0 +1,24 @@
+  "trigger_time_inf_freq_error": 0.013108099559195993,
+  "importance": 0.9871308604784662,
+  "sampling_time": 0.983,
+  "spectrum": "wavelength",
```
Could you please update according: https://github.com/nasa-gcn/gcn-schema/blob/main/gcn/notices/core/Reporter.schema.json?
`spectrum` and `units` were redundant, later only `units` is kept in PR: https://github.com/nasa-gcn/gcn-schema/pull/248

Vidushi-GitHub replied:
@tabbott36 please check it.

tabbott36 replied:
Removed spectrum from the examples jsons, thanks!

jracusin replied:
We're changing the core schema reporter.units from Hz to MHz.  Use spectral_band and units.

jracusin commented on gcn/notices/chime/frb.detection.example.json:
```diff
@@ -0,0 +1,23 @@
+  "importance": 0.9871308604784662,
+  "sampling_time": 0.983,
+  "spectral_center": 600,
+  "bandwidth": 400,
```
We agreed to switch to `spectral_band` and `units`, no `bandwidth`.

Vidushi-GitHub commented on gcn/notices/chime/frb.detection.example.json:
```diff
@@ -0,0 +1,23 @@
+  "alert_type": "initial",
+  "trigger_time": "2024-09-18T07:19:10.765268Z",
+  "trigger_time_error": 0.00786431,
+  "id": ["427325191"],
```
Hi, could you please update it as only single string, array is not must anymore for single IDs?
See: https://github.com/nasa-gcn/gcn-schema/blob/main/gcn/notices/core/Event.schema.json

Vidushi-GitHub reviewed (approved):
Please use `id` as single string for one ID, instead of array, we upgraded the core.
Then squash and rebase.
It's ready to merge after that.

Vidushi-GitHub commented on .github/workflows/release.yml:
```diff
@@ -17,7 +17,7 @@ jobs:
     steps:
       - uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683 # v4.2.2
-      - uses: actions/setup-node@cdca7365b2dadb8aad0a33bc7601856ffabcc48e # v4.3.0
+      - uses: actions/setup-node@49933ea5288caeca8642d1e84afbd3f7d6820020 # v4.4.0
```
@tabbott36 please reverse this change.

tabbott36 replied:
I may be wrong, but I believe these changes are a result of running `npm install` which was necessary to pass checks.

Vidushi-GitHub commented on .github/workflows/validate.yml:
```diff
@@ -17,7 +17,7 @@ jobs:
       - uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683 # v4.2.2
       - name: Install Node.js
-        uses: actions/setup-node@cdca7365b2dadb8aad0a33bc7601856ffabcc48e # v4.3.0
+        uses: actions/setup-node@49933ea5288caeca8642d1e84afbd3f7d6820020 # v4.4.0
```
Reverse it

Vidushi-GitHub reviewed (changes requested):
@tabbott36 by mistakenly unrelated to CHIME schema files are touched.
Please reverse those changes.

Vidushi-GitHub commented:
@dakota002 is this looks final in terms of squash and rebase?
Please help @tabbott36 if you have suggestion.

tabbott36 commented:
> @dakota002 is this look fine in terms of squash and rebase? Please help @tabbott36 if you have suggestion.

I'm happy with it! The only issue I am aware of is that it's not passing tests which I haven't been able to debug since `npm run test` does not return any errors.

dakota002 commented:
> > @dakota002 is this look fine in terms of squash and rebase? Please help @tabbott36 if you have suggestion.
> 
> I'm happy with it! The only issue I am aware of is that it's not passing tests which I haven't been able to debug since `npm run test` does not return any errors.

Hey @tabbott36, it looks like it is just an issue with the formatting checks. Please run `npx prettier -w .` locally, then try pushing out your changes

tabbott36 commented:
Thanks @dakota002, that did it!

Vidushi-GitHub commented:
@tabbott36 looks good, thanks.
Could you please rebase and squash the commits?
